tests/packages/sharedlib-in-package/mypkg/__init__.py

```diff
@@ -0,0 +1,25 @@
+# SPDX-FileCopyrightText: 2024 The meson-python developers
+#
+# SPDX-License-Identifier: MIT
+
+import os
+
+
+def _load_sharedlib():
+    """Load the `example_lib.dll` shared library on Windows
+
+    This shared library is installed alongside this __init__.py file. Due to
+    lack of rpath support, Windows cannot find shared libraries installed
+    within wheels. So pre-load it.
+    """
+    if os.name == "nt":
+        from ctypes import WinDLL
+        basedir = os.path.dirname(__file__)
+        dll_path = os.path.join(basedir, "example_lib.dll")
+        WinDLL(dll_path)
+
+
+_load_sharedlib()
```

Comment:
Isn't `os.add_dll_directory()` the recommended way to deal with this? https://docs.python.org/3/library/os.html#os.add_dll_directory

Reply:
That's true now, yes. The pain point is that it can only be relied on for Python >=3.10. It wasn't useful until the documented changes in 3.8, and those took 2 more minor versions to propagate to at least Conda's `python` builds.

We could use `os.add_dll_directory` in the docs here, and note that if a package still supports <3.10, they should use `ctypes.WinDLL` preloading instead?

Reply:
We're also still testing with Python 3.7, which is EOL. Maybe time to drop that for the next release?

Reply:
I'm not sure I'm following. The documentation states that this function was added in Python 3.8. Are you saying that Conda's Python carried some patches that broke this and it was fixed only in Conda's Python 3.10?

The `ctypes.WinDLL` does not work either, judging from the results of CI 😄

Reply:
Oh, but the CI issue is just a typo in the name of the library that is being loaded.

Reply:
Yes `WinDLL` should work, since we're using it in SciPy.

And re Conda: I remember it was a long story, but the tl;dr I took from it was that CPython upstream's changes were quite disruptive (not surprising, since Conda is the only real distro on Windows, and had already needed to implement its own DLL loading rules/precedence) and it took two years for Conda to adapt.

Reply:
> We're also still testing with Python 3.7, which is EOL. Maybe time to drop that for the next release?

So far it would have required more effort to drop support for Python 3.7 than to keep it, so it hasn't happened. If there is any reason for dropping it (including support from the CI infrastructure) I don't see any reason why we should keep it.

Reply:
I think mentioning that the `os.add_dll_directory()` thing does not work on Conda's Python before 3.10. Not knowing better, I would have used it and debugging while it does not work on Conda would have not been fun.

Reply:
Yeah it wasn't at the time, we had a bad time with wheels vendoring shared libs not working with Conda's Python. 

I'll finish debugging this new test on my own fork or find a Windows machine, enough abusing CI here.